src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/MatchXpathCheck.java

```diff
@@ -0,0 +1,274 @@
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2020 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+import com.puppycrawl.tools.checkstyle.xpath.AbstractNode;
+import com.puppycrawl.tools.checkstyle.xpath.RootNode;
+import net.sf.saxon.Configuration;
+import net.sf.saxon.om.Item;
+import net.sf.saxon.sxpath.XPathDynamicContext;
+import net.sf.saxon.sxpath.XPathEvaluator;
+import net.sf.saxon.sxpath.XPathExpression;
+import net.sf.saxon.trans.XPathException;
+
+/**
+ * <p>
+ * Evaluates Xpath query and report violation on all matching AST nodes. Current check allows
+ * user to implement custom checks using Xpath. If Xpath query is not specified explicitly,
+ * then check does nothing.
+ * </p>
+ * <p>
+ * Please read more about Xpath syntax at <a href="https://www.w3schools.com/xml/xpath_syntax.asp">
+ * W3Schools Xpath Syntax</a>. Information regarding Xpath functions can be found at
+ * <a href="https://developer.mozilla.org/en-US/docs/Web/XPath/Functions">XSLT/XPath Reference</a>.
+ * Note, that <b>@text</b> attribute can used only with token types that are listed in
+ * <a href="https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/utils/XpathUtil.java#L101">
+ *     XpathUtil</a>.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code query} - Xpath query.
+ * Type is {@code java.lang.String}.
+ * Default value is {@code ""}.
+ * </li>
+ * </ul>
+ * <p>
+ * Checkstyle provides own cmd tool and here the basic options will be
+ * shown to ease usage of Xpath queries. More options and detailed information can be found
+ * at <a href="https://checkstyle.org/cmdline.html">Command Line</a>.
+ * </p>
+ * <p><b>-t</b> option prints AST tree of the checked file.</p>
```

Comment:
`-T`, as we are about to make it comments also.
if you want to keep `-t`, please add one more acceptance criteria to #9009

Reply:
done